### PR TITLE
Javascript tests

### PR DESCRIPTION
**Pull request by jklingen92**
Addresses Issue #3: Javascript Tests

This PR contains javascript tests for the default wrappers. More tests are on their way, but y'all can take a look and see if they're ok so far. Looking for feedback on form and style.

Note: This PR currently includes changes from PR #18 until that gets merged in. 

EDIT: Added test for widget_form and utils, currently testing at 95% (94.95% of lines technically).

### CODE REVIEW

**Comment by noisecapella** [comment]
Can you update `travis/js_tests.sh` to run the tests on travis?

**Comment by noisecapella** [comment]
Since #18 is merged can you rebase please?

File: open-widget-framework/test/defaults_test.js [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+  })
+
+  // Method + render: Ensure that method calls cause the appropriate components to be rendered
+  it('calls renderList and no other form render functions by default', () => {
+    resetSpyHistory()
+    mount(<ListWrapper {...dummyProps}/>)
+
+    expect(dummyProps.renderNewWidgetForm.callCount).to.equal(0)
+    expect(dummyProps.renderEditWidgetForm.callCount).to.equal(0)
+    expect(dummyProps.renderList.callCount).to.equal(1)
```
Comment (noisecapella) [inline open-widget-framework/test/defaults_test.js]: Can we also assert that the return of `renderList` appears in the rendered output?

File: open-widget-framework/test/defaults_test.js [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+  })
+
+  it('addWidget causes just renderNewWidgetForm to be called', () => {
+    const wrap = mount(<ListWrapper {...dummyProps}/>)
```
Comment (noisecapella) [inline open-widget-framework/test/defaults_test.js]: Can we also assert that the return value of `renderNewWidgetForm` appears in the rendered output?

File: open-widget-framework/test/defaults_test.js [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+    resetSpyHistory()
+    instance.addWidget()
+
+    expect(dummyProps.renderNewWidgetForm.callCount).to.equal(1)
+    expect(dummyProps.renderEditWidgetForm.callCount).to.equal(0)
+    expect(dummyProps.renderList.callCount).to.equal(1)
+  })
+
+  it('editWidget causes just renderEditWidgetForm to be called', () => {
+    const wrap = mount(<ListWrapper {...dummyProps}/>)
```
Comment (noisecapella) [inline open-widget-framework/test/defaults_test.js]: Can we also assert that the return value of `renderEditWidgetForm` appears in the rendered output?

File: open-widget-framework/test/defaults_test.js [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+  })
+
+  it('toggleEditWidget causes renderAddWidgetButton to be called', () => {
+    const wrap = mount(<ListWrapper {...dummyProps}/>)
+    const instance = wrap.instance()
+    resetSpyHistory()
+    const renderAddWidgetButtonSpy = sinon.spy(instance, 'renderAddWidgetButton')
+    instance.toggleEditMode()
+
+    expect(renderAddWidgetButtonSpy.callCount).to.equal(1)
```
Comment (noisecapella) [inline open-widget-framework/test/defaults_test.js]: Can we also assert that the output of `renderAddWidgetButton` appears in the `wrap` rendered output?

File: open-widget-framework/test/defaults_test.js [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+    title: dummyTitle,
+    html: 'dummyHTML',
+  }
+
+  it('sets title and html elements on render', () => {
+    const wrap = mount(<Renderer {...dummyProps}/>)
+    const htmlWrap = mount(dummyHTML)
+
+    expect(wrap.find('.widget-title').text()).to.equal(dummyTitle)
+    expect(wrap.find('.widget-text').text()).to.equal(htmlWrap.text())
```
Comment (noisecapella) [inline open-widget-framework/test/defaults_test.js]: Can we change this to compare the HTML instead of the text? We should also change the `html` prop in `dummyProps` to something containing HTML so we can make sure it is being rendered correctly.

File: open-widget-framework/test/defaults_test.js [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+
+describe('<_defaultFormWrapper />', () => {
+  const dummyData = {data: 'dummyData'}
+  const dummyProps = {
+    updateWidgetList: sinon.spy(),
+    renderForm: sinon.spy(),
+    closeForm: sinon.spy(),
+  }
+
+  const resetSpyHistory = () => {
```
[the remark this answers is not in the document]
Comment (jklingen92) [reply open-widget-framework/test/defaults_test.js]: Not exactly. When the component is mounted, some of the functions run during the rendering. The callCounts I'm checking for assume that all callCounts are zero and then one single helper function is called. I could additionally put it in an afterEach, but that wouldn't let me remove it from the tests without reassessing all the tests that fail (11 of them I believe)

File: open-widget-framework/test/defaults_test.js (hunk 1 above)
Comment (jklingen92) [reply open-widget-framework/test/defaults_test.js]: `renderList` is a prop that is opaque to the wrapper function. In the tests it's just a spy function so it won't have any returns. `widget_list_test.js` tests all of the `WidgetList` methods which are the functions that are eventually passed into the default wrappers

File: yarn.lock [hunk 7]
```diff
@@ -0,0 +1,4 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
[the remark this answers is not in the document]
Comment (jklingen92) [reply yarn.lock]: By mistake

File: open-widget-framework/test/defaults_test.js (hunk 5 above)
Comment (jklingen92) [reply open-widget-framework/test/defaults_test.js]: I put the whole expected div in as `dummyHTML`. We should definitely have HTML in the `html` prop.

File: open-widget-framework/yarn.lock [hunk 8]
```diff
@@ -29,7 +29,7 @@
     semver "^5.4.1"
```
[the remark this answers is not in the document]
Comment (jklingen92) [reply open-widget-framework/yarn.lock]: Nope. I haven't used yarn before, but I'm adding the lock file to the git ignore.

File: open-widget-framework/package.json [hunk 9]
```diff
@@ -15,7 +15,7 @@
     "clean": "nwb clean-module && nwb clean-demo",
     "prepublishOnly": "npm run build",
     "start": "nwb serve-react-demo",
-    "test": "nwb test-react",
+    "test": "nyc mocha",
     "test:coverage": "nwb test-react --coverage",
     "test:watch": "nwb test-react --server"
```
[the remark this answers is not in the document]
Comment (jklingen92) [reply open-widget-framework/package.json]: What are you referring to?

File: open-widget-framework/test/defaults_test.js [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+  _defaultRenderer as Renderer,
+  _defaultLoader as Loader,
+} from '../src/defaults'
+
+describe('_defaultFetchJsonData', () => {
+  const dummyUrl = "some.dummy.url"
+  const dummyPostInit = {
+    method: 'POST',
+  }
+  // TODO: implement
```
[the remark this answers is not in the document]
Comment (noisecapella) [reply open-widget-framework/test/defaults_test.js]: It looks like it's possible to set `global.fetch` to a sinon stub which would allow you to mock the built-in fetch.

File: open-widget-framework/package.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (noisecapella) [reply open-widget-framework/package.json]: The `nwb test-react --server` command here does not show any tests being run. In the command above `nwb test-react` was replaced with `nyc mocha` but this line has not changed, so I think we need to either change this line or remove this command if we don't need watch functionality.

File: open-widget-framework/test/widget_list_test.js [hunk 11]
```diff
@@ -1,31 +1,324 @@
...
+    const wrap = mount(<WidgetList {...dummyProps}/>)
+    const instance = wrap.instance()
+    instance.updateWidgetList(dummyWidgetInstances)
+    const dummyWidgetInstance = dummyWidgetInstances[0]
+    const dummyWidgetId = dummyWidgetInstance.id
+
+    const widgetWrap = mount(instance.renderWidget(dummyWidgetInstance, dummyProps.listWrapperProps))
+    expect(widgetWrap.exists('#widget-' + dummyWidgetId)).to.equal(true)
+    expect(widgetWrap.props()).to.include(dummyWidgetInstance)
+    // TODO: figure out how to check for the passthrough props
```
Comment (noisecapella) [inline open-widget-framework/test/widget_list_test.js]: Can this be done similar to the `renderWidgetList renders a ListWrapper` test above?

File: open-widget-framework/test/widget_list_test.js [hunk 12]
```diff
@@ -1,31 +1,324 @@
...
+    const listWrap = mount(instance.renderWidgetList())
+    expect(listWrap.exists('.default-list-wrapper')).to.equal(true)
+    expect(listWrap.props()).to.include(instance.makePassThroughProps())
+    expect(listWrap.props()).to.include(dummyProps.listWrapperProps)
+  })
+
+  it('renderListBody runs render widget once for each widget in widget instances', () => {
+    const wrap = mount(<WidgetList {...dummyProps}/>)
+    const instance = wrap.instance()
+    instance.updateWidgetList(dummyWidgetInstances)
```
Comment (noisecapella) [inline open-widget-framework/test/widget_list_test.js]: Can you add a test where `updateWidgetList` is not called before running `renderListBody`? When testing via the sample app I noticed that the `widgetInstances` state is sometimes null. Is it possible to change this to never be null, or otherwise to make sure we're checking null status in each function?

File: open-widget-framework/test/defaults_test.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (jklingen92) [reply open-widget-framework/test/defaults_test.js]: Added node-fetch and stubbed that

File: open-widget-framework/test/widget_form_test.js [hunk 13]
```diff
@@ -0,0 +1,338 @@
...
+
+describe('<NewWidgetForm />', () => {
+  // Test default behavior
+  it('returns loader if no data is loaded', () => {
+    const wrap = mount(<NewWidgetForm {...dummyNewWidgetFormProps}/>)
+
+    expect(wrap.exists('.default-loader')).to.equal(true)
+  })
+
+  it('renders a blank WidgetForm if data is loaded and that form submits properly', () => {
```
Comment (noisecapella) [inline open-widget-framework/test/widget_form_test.js]: Can we add a test where `onSubmit` gets a null value for `formData`? I ran into a null exception in the `onSubmit` handler here when I created a text input then clicked 'submit'.

File: open-widget-framework/test/widget_list_test.js (hunk 12 above)
Comment (jklingen92) [reply open-widget-framework/test/widget_list_test.js]: The reason we're using `null` is to be able to differentiate between waiting for the async data to be returned and being given an empty list. We could use `undefined` if that's any better. I worry that if we set it to an empty array, it will muck with the the wrapper and may not refresh when the async function does return (for example, we would have to run makePassThroughProps again when the data returns or the props passed to ListWrapper are inaccurate).

File: travis/js_tests.sh [hunk 14]
```diff
@@ -4,5 +4,8 @@ cd open-widget-framework
```
Comment (noisecapella) [inline travis/js_tests.sh]: On the second line of this file can you add:

    set -e -o pipefail

This would cause the file to return a non-zero exit value  if any of the commands in the file return a non-zero exit file

File: open-widget-framework/test/widget_form_test.js [hunk 15]
```diff
@@ -0,0 +1,338 @@
...
+    })
+
+    const instance = wrap.instance()
+    instance.onSubmit(document.createEvent('Event'))
+    expect(wrap.prop('onSubmit').withArgs(dummyWidgetClass, dummyFormData).callCount).to.equal(1)
+    resetSpyHistory()
+  })
+
+  it('makeWidgetClassSelect creates a select with options that match the widgetClasses given', () => {
+    const wrap = mount(<WidgetForm {...dummyNewFormProps}/>)
```
[the remark this answers is not in the document]
Comment (jklingen92) [reply open-widget-framework/test/widget_form_test.js]: This is in the new version

File: open-widget-framework/test/widget_list_test.js (hunk 11 above)
Comment (jklingen92) [reply open-widget-framework/test/widget_list_test.js]: I was able to check for the non-widget specific props and then check for the presence of the four widget specific props. The makePassThroughProps test checks that those props are set to appropriate values by the function so that should cover everything.

File: open-widget-framework/test/widget_list_test.js (hunk 12 above)
Comment (jklingen92) [reply open-widget-framework/test/widget_list_test.js]: @noisecapella This is the last remaining issue on the PR

Comment (noisecapella) [reply open-widget-framework/test/widget_list_test.js]: I'm fine with using `null` in that way, but there should be tests for `updateWidgetList` where `widgetInstances` is null so we can verify that the case is handled correctly.
